cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/preview/DefaultPreviewRunner.java

```diff
@@ -252,6 +257,7 @@ protected void startUp() throws Exception {
       ((Service) messagingService).startAndWait();
     }
     datasetService.startAndWait();
+    StoreDefinition.createAllTables(structuredTableAdmin, true);
```

Comment:
hmmm. The fact that you pass in overwrite=true tells me that this would fail because the tables already exist. Does that mean that preview's tables will clash with the CDAP master's tables? I am not that familiar with how preview sets up its datasets. But I believe that it needs its own instance of the system tables. Can you review this with @albertshau ?

Reply:
For the SQL case - we'll need to pass in another dataSource in for preview. If we can have two EmbeddedPostgres' running at once - it should be fine. This will depend on how exactly the injection will be handled for Preview. I think as long as dataSource is not singleton - this should work.

Reply:
Shouldn't the preview always use the local (leveldb) implementation?

Reply:
That also can be configured - it will depend on how injection is done for nosql vs sql. We can have the preview runner injector always use nosql if that's what we want.

Reply:
Any system table or dataset created in preview space is supposed to be nosql leveldb, regardless of what the CDAP master is configured to use. I would expect there to be extra work here in setting up the bindings, and that this would actually be easier to implement if preview were isolated in its own service instead of sharing the app-fabric bindings. However, it seems like that could be done in a separate PR.

Regarding the override flag, I took a brief look at this yesterday and the problem seems to be that table existence is handled differently now with the new SPI. There is some StructuredTableSpecificationRegistry that just uses an in-memory map to decide which StructuredTables exist. That same instance is being used in preview space, so the 'namespaces' StructuredTable comes back as existing even though it actually doesn't. So "overriding" the table doesn't actually override anything because it doesn't actually exist underneath.

I believe the right way to fix the problem is to make sure that the preview run gets a new instance of StructuredTableSpecificationRegistry. That class looks odd and is documented as temporary, so this might not be needed whenever it becomes a real implementation.

Reply:
Thanks Albert. We eventually need to migrate the registry to an actual dataset, so thats when I would migrate everything to do things the correct way. Since both are temporary - I'd prefer to get this in now and migrate other datasets.

Reply:
Won't you still need a separate instance in that case? I don't think overwrite is ever the correct solution, now or in the future.

Reply:
No, the point of the overwrite is to ensure we create a system entity table in preview space. We don't really care about the schema map in preview vs non preview space since all system datasets will have the same schema anyways.

Reply:
The only reason you have to set overwrite is because the system thinks it already exists when it doesn't. It's correct to create all the system tables in preview, but it's not correct to overwrite. It's not correct that it thinks it exists.

Forcing your way past incorrect behavior here could cover up some actual problem that occurs somewhere else in the code.

Reply:
Whats the difference in behavior of overwrite vs passing in a new map for preview after initialization? In both cases, in preview space, we will see that we have initialized all system datasets, and in normal space, the same is true. Regardless if we overwrite or pass in a new map, we get the same configuration for preview and normal space after initialization. Since we're getting rid of the map anyways - I don't see the harm in doing this. 

The incorrectness that you mentioned is only valid when we haven't created any tables in preview. Since this function is called when preview starts up - after the call there is no difference.

Reply:
The harm is that you're relying on a bug to accomplish something instead of going through the proper channel. This is coupled very specifically to how the storage SPI is implemented several abstraction layers below.

Reply:
After discussing this more in person, we decided the registry changes will happen later in a future PR, but we'll mark those changes as a blocker. Here we'll just put a comment about why the overwrite flag is needed, with a TODO to remove it.